tests/test_pipeline.py

```diff
@@ -0,0 +1,446 @@
+            # constraints is:
+            #  grad_objective + grad_constraint * relu(multiplier + penalty_coeff * violation)
+            violation = self._manual_violation(manual_x)[features]
+            aux_grad = manual_multiplier[features] + manual_penalty_coeff[features] * violation
+            if self.is_inequality:
+                aux_grad.relu_()
+            manual_x_grad = 2 * manual_x + lhs[features].t() @ aux_grad
+        else:
+            manual_x_grad = 2 * manual_x + lhs[features].t() @ manual_multiplier[features]
+        manual_x = manual_x - PRIMAL_LR * manual_x_grad
+        return manual_x
+
+    def _manual_dual_step(self, manual_x, manual_multiplier, strict_features, manual_penalty_coeff=None):
+        violation = self._manual_violation(manual_x, strict=True)[strict_features]
+        if manual_penalty_coeff is None:
+            manual_multiplier[strict_features] = manual_multiplier[strict_features] + DUAL_LR * violation
+        else:
+            manual_multiplier[strict_features] = (
+                manual_multiplier[strict_features] + manual_penalty_coeff[strict_features] * violation
+            )
+
+        if self.is_inequality:
+            manual_multiplier.relu_()
+        return manual_multiplier
+
+    def _manual_primal_lagrangian(
+        self, manual_x, manual_multiplier, features, manual_penalty_coeff: Optional[torch.Tensor] = None
+    ):
+        violation = self._manual_violation(manual_x)[features]
+        loss = torch.sum(manual_x**2)
+
+        if manual_penalty_coeff is None:
+            return loss + torch.dot(manual_multiplier[features], violation)
+
+        penalty_term = manual_multiplier[features] + manual_penalty_coeff[features] * violation
+        if self.is_inequality:
+            penalty_term.relu_()
+        aug_lag = loss + 0.5 * torch.dot(
+            1 / manual_penalty_coeff[features], penalty_term**2 - manual_multiplier[features] ** 2
+        )
+        return aug_lag
+
+    def _manual_dual_lagrangian(self, manual_x, manual_multiplier, strict_features, manual_penalty_coeff=None):
+        violation = self._manual_violation(manual_x, strict=True)[strict_features]
+        if manual_penalty_coeff is None:
+            return torch.sum(manual_multiplier[strict_features] * violation)
+        return torch.sum(manual_penalty_coeff[strict_features] * manual_multiplier[strict_features] * violation)
+
+    def _manual_simultaneous_roll(self, manual_x, manual_multiplier, features, strict_features):
+        manual_primal_lagrangian = self._manual_primal_lagrangian(manual_x, manual_multiplier, features)
+        manual_observed_multipliers = manual_multiplier.clone()
+        manual_dual_lagrangian = self._manual_dual_lagrangian(manual_x, manual_multiplier, strict_features)
+        manual_x, manual_multiplier = self._manual_primal_step(
+            manual_x, manual_multiplier, features
+        ), self._manual_dual_step(manual_x, manual_multiplier, strict_features)
```

Comment:
Is this not split into two calls to ensure "simultaneous" updates?
Can we make it fit in one line then? By defining kwargs. Also, it might be worth documenting this fact

Reply:
Both updates will still not fit in one line